File: typescript/src/utils/progressSerializer.ts
import type { UploadProgress } from '../types/index.js';

export function serializeProgress(progress: UploadProgress): string {
  const data = {
    processedIds: Array.from(progress.processedIds),
    lastProcessedRow: progress.lastProcessedRow,
    failedUploads: progress.failedUploads,
  };
  return JSON.stringify(data, null, 2);
}

export function deserializeProgress(data: string): UploadProgress {
  try {
    const parsed = JSON.parse(data);
    return {
      processedIds: new Set(parsed.processedIds ?? []),
      lastProcessedRow: parsed.lastProcessedRow ?? 0,
      failedUploads: parsed.failedUploads ?? [],
    };
  } catch {
    return {
      processedIds: new Set(),
      lastProcessedRow: 0,
      failedUploads: [],
    };
  }
}

File: python/src/utils/progress_serializer.py
"""Serialize and deserialize progress data."""

import json
from typing import Any, Dict

from models import UploadProgress


def serialize_progress(progress: UploadProgress) -> str:
    """
    Convert UploadProgress object to JSON string for persistence.

    Args:
        progress: UploadProgress object to serialize

    Returns:
        Pretty-printed JSON string
    """
    return json.dumps(progress.to_dict(), indent=2)


def deserialize_progress(data: str) -> UploadProgress:
    """
    Parse JSON string into UploadProgress object.

    Args:
        data: JSON string to parse

    Returns:
        UploadProgress object (empty if parse fails)
    """
    try:
        parsed_data = json.loads(data)
        return UploadProgress.from_dict(parsed_data)
    except (json.JSONDecodeError, KeyError, TypeError):
        # Return empty progress on any parse failure
        return UploadProgress()
